Adding the Bots documentation (#77)
ONSdigital/dp-developer-site: static/bots/index.md

---
title: Guide to rate limiting and bot development
---

When creating a bot, scraper or automation to programmatically get data from the ONS site and APIs, please read the following guidance.

## Rate limiting

The ONS dissemination sites and API apply rate limiting to ensure a high-quality service is delivered to all users.

The following rate limits have been applied:

* 120 requests per 10 seconds - all site and API assets

* 200 requests per 1 minute - all site and API assets

* 15 requests per 10 seconds - high demand site assets

The time for which you will be blocked varies and is dynamically changed by the ONS, depending on release schedules and observed demand.

If you exceed these limits, the API will return a `429 Too Many Requests` HTTP status code and a `Retry-After` header containing the number of seconds until you may try your request again.

> The ONS expects that you programmatically respect the Retry-After header and back-off your polling accordingly. If this is not respected, our algorithms will automatically impose a total block to our services for anything up to 24 hours.
> 
> Your tooling should respect the `Retry-After` header and pause your requests for the specified duration. If this is not respected our algorithms may impose a block to our services for up to 1 hour.

Reviewer: Looks like we have both versions here still:

```suggestion
Your tooling should respect the `Retry-After` header and pause your requests for the specified duration. If this is not respected our algorithms may impose a block to our services for up to 1 hour.
```

> Read more in our [fair usage policy](TODO)

Reviewer: We should probably include something to say that we reserve the right to change these at any time. Perhaps this paragraph is the correct place?

```suggestion
We reserve the right to change our rate limits at any time and without prior notice. Read more in our [fair usage policy](TODO)
```

Reviewer: Also this URL will be based on the page title (which is actually going to be `fair use policy`) so will be generated as:

```suggestion
Read more in our [fair use policy](https://www.ons.gov.uk/help/fairusepolicy).
```
Also added missing full stop.

### Naming your bot and versioning

Naming your bot and providing a version makes it easier for us to identify and work with you should there be any problems.

```text
botName/Version1.0.0
```

### Include a URL or contact information in your user-agent

When writing bots, it is best practice to include a means of contacting the bot owner if necessary. The best method is to include a URL to a page detailing the owner of the bot, the purpose of the bot and providing a means of contacting you. Alternatively, you can include a company email address (do not use a personal email) we can use to contact you for further information.

Examples:

```text
botName/Version1.0.0 (organisation-name contact@organisation.com +http://www.organisation-site.com/bot.html)
botName/Version1.0.0 (organisation-name +http://www.organisation-site.com/bot.html)
botName/Version1.0.0 (organisation-name contact@organisation.com)
```

### Do not use personally identifying information in the `User-Agent` header

Do not include your name, username or personal email in the `User-Agent`. Your company name is acceptable if you have one.

> Anything you provide will be stored in our logs as per our [data retention policy](https://www.ons.gov.uk/aboutus/transparencyandgovernance/dataprotection/dataprotectionpolicy)

Reviewer: I was just reviewing our policies and I think this should actually refer to our T&Cs instead (privacy and cookies section):

```suggestion
Anything you provide will be stored in our logs as per our [terms and conditions](https://www.ons.gov.uk/help/termsandconditions).
```
(note the missing full stop as well)

---
title: Bots
---

> Hello world 3

Reviewer 2: Hello world 3?

Author: I was testing the dynamic reload... removed

> When creating a bot, scrapper or automation to programmatically get data from the ONS site and APIs please read the following guidance.

Reviewer 2: *scraper

## Rate limiting

> The ONS API applies rate limiting to ensure a high quality service is delivered to all users and to protect client applications from unexpected loops.

Reviewer 2: ```suggestion
The ONS website and API apply rate limiting to ensure a high quality service is delivered to all users and to protect client applications from unexpected loops.
```

> The ONS API applies rate limiting to ensure a high quality service is delivered to all users and to protect client applications from unexpected loops.

Author: The loops bit was replicated from the original copy, will remove 👍

* 120 requests per 10 seconds - all site and API assets

* 200 requests per 1 minute - all site and API assets

* 15 requests per 10 seconds - high demand site assets

The time for which you will be blocked varies and is dynamically changed by the ONS, depending on release schedules and observed demand.

If you exceed these limits, the API will return a `429 Too Many Requests` HTTP status code and a `Retry-After` header containing the number of seconds until you may try your request again.

> The ONS expects that you programmatically respect the Retry-After header and back-off your polling accordingly. If this is not respected, our algorithms will automatically impose a total block to our services for anything up to 24 hours.

Reviewer: I think we could soften the language a bit on this and 24h may be too aggressive. I would suggest we keep this as 1 hour?

```suggestion
Your tooling should respect the `Retry-After` header and pause your requests for the specified duration. If this is not respected our algorithms may impose a block to our services for up to 1 hour.
```

Author: Amelia approved 👍

ONS expect that you programmatically respect the Retry-After header and back off your polling accordingly. If this is not respected our algorithms will automatically impose a total block to our services for anything up to 24hrs.

## Constructing a good user-agent for Bot usage

The `user-agent` header informs us of the characteristics of the requesting user agent.

### Naming your Bot and versioning

Naming your bot and providing a version makes it easier for us to identify and work with you should there be any problems.

```shell
botName/Version1.0.0
```

### Don't leak personally identifying information in the user agent

> Do not include your name, username or personal email in the user-agent. It is not necessary to provide us such information and not the intention of the `user-agent` header.
> 
> Your company name is acceptable if you have one.

Reviewer: I think we can shorten this a bit as well:

```suggestion
Do not include your name, username or personal email in the user-agent. Your company name is acceptable if you have one.
```

Anything you provide will be stored in our logs as per our [data retention policy](https://www.ons.gov.uk/aboutus/transparencyandgovernance/dataprotection/dataprotectionpolicy)

> ### If you're writing a bot, make sure to include some kind of url/contact info

Author: Currently changed to Amelia's suggestion
`If you are writing a bot include a URL or contact information`

> This can be an company email address (not a personal one) or a url to a page you host to define your bots intentions or contact information:

Reviewer 2: ```suggestion
This can be an company email address (not a personal one) or a url to a page you host to define your bot's intentions or contact information:
```

Just a URL

```shell
botName/Version1.0.0 (organisation-name +http://www.organisation-site.com/bot.html)
```

Just contact information

```shell
botName/Version1.0.0 (organisation-name contact@organisation.com)
```

> ### Do not use personally identifying information in the `user-agent` header
> 
> Do not include your name, username or personal email in the `user-agent`. Your company name is acceptable if you have one.

Reviewer: I would normally expect header names to be Title Case:

```suggestion
### Do not use personally identifying information in the `User-Agent` header

Do not include your name, username or personal email in the `User-Agent`. Your company name is acceptable if you have one.
```

botName/Version1.0.0 (organisation-name contact@organisation.com)
```

> ### Don't use HTML tags

Reviewer: ```suggestion
### Do not use HTML tags
```

When providing a URL as your bot information please refrain from using the HTML tags as follows:

```shell

> botName/Version1.0.0 (organisation-name <a href="http://www.organisation-site.com/bot.html>My Bot</a>)

Reviewer: Missing closing quote is breaking formatting:

```suggestion
botName/Version1.0.0 (organisation-name <a href="http://www.organisation-site.com/bot.html">My Bot</a>)
```

```

Instead use;

```shell
botName/Version1.0.0 (organisation-name +http://www.organisation-site.com/bot.html)
```

### Keep it simple no UTF8 or emojis

> We have no problems processing UTF8 or even emojis, but its not necessary, it's a better practice to keep your `user-agent`` as simple as possible - only use letters, numbers and basic symbols (brackets, hypens, slashes etc).

Reviewer 2: ```suggestion
We have no problems processing UTF8 or even emojis, but its not necessary, it's a better practice to keep your `user-agent` as simple as possible - only use letters, numbers and basic symbols (brackets, hypens, slashes etc).
```
